README
=========================================
README - pyside-assistant generated files
=========================================

All files generated by psa init are described below. The sampleproject
slug used for them was 'sampleproject', with the following psa call:

    $ psa init sampleproject <fremantle|harmattan>

* MANIFEST.in: Distutils considers a list of files by default for
installation, listed at [1].  Additional files to be distributed
must be supplied in the MANIFEST.in file. The generated file includes
the .desktop file and any QML files found.

[1] http://docs.python.org/distutils/sourcedist.html#specifying-the-files-to-distribute.

* setup.py: The main file for using Distutils, contains most of the
information needed to build the package. Information about setup.py
can be found at http://docs.python.org/distutils/introduction.html.
Two fields of this file can be modified by psa update parameters:
 - description: -d or --description
 - long_description: the contents of sampleproject.longdesc are used

* stdeb.cfg: Configuration file for stdeb, specifying additional
parameters for the binary package. The specification for this
file can be found at https://github.com/astraw/stdeb, in the section
stdeb.cfg configuration file. One field of this file can be modified by
psa update parameters:
 - Section: -s or --section

Note that the section should be a valid one,
as specified in http://wiki.maemo.org/Task:Package_categories#New_list_for_Diablo.

* sampleproject.aegis: Sample credentials file, which initially is empty. Information
of how to populate it can be found at [2].
NOTE: this file is used only in harmattan projects.

[2] http://library.developer.nokia.com/topic/MeeGo_1.2_Harmattan_API/html/guide/html/Developer_Library_Developing_for_Harmattan_Harmattan_security_6cbe.html

* sampleproject.desktop: This file specifies how to run the application
from the application grid, with various fields. Information about
.desktop files can be found at http://wiki.maemo.org/Desktop_file_format.
Two fields of this file can be modified by psa update parameters:
 - Name: -a or --app-name
 - Category: -c or --category

Note that the category should be a valid one, as specified in
http://standards.freedesktop.org/menu-spec/latest/apa.html.

* sampleproject.png: The application icon, which is displayed in the
application grid.

* qml/*.qml: The QML files for the application. Their contents depend
on the platform.

* sampleproject: Main program. Initializes the application and provide
support for displaying the QML files contents.

* sampleproject.longdesc: Holds the contents of the long_description field of
setup.py, which as the name implies is a more detailed description of what the project is.



debian/changelog
pyside-assistant (0.6) unstable; urgency=low

  * Fix bug #978 - Unexpected dependency for Python 2.7 using pyside-assistant
  * Add build dep: quilt
  * New patch:
    + 01-ubuntu-oneiric-fix.patch

 -- Bruno Araujo <bruno.araujo@openbossa.org>  Wed, 19 Oct 2011 13:57:17 -0400

pyside-assistant (0.5-ubuntu4~n1) natty; urgency=low

  * Refactor
  * Split templates from commands

 -- Lauro Moura <lauro.neto@openbossa.org>  Thu, 04 Aug 2011 17:44:36 -0300

pyside-assistant (0.4) unstable; urgency=low

  [ Willer Moreira ]
  * Adding support to fremantle

  [ Bruno Araujo ]
  * Update unit tests
  * Detect python2.5 when creating Fremantle projects

 -- Bruno Araujo <bruno.araujo@openbossa.org>  Mon, 04 Jul 2011 14:33:34 -0400

pyside-assistant (0.3) unstable; urgency=low

  * Fix init command options validation

 -- Bruno Araujo <bruno.araujo@openbossa.org>  Wed, 29 Jun 2011 14:35:15 -0400

pyside-assistant (0.2) unstable; urgency=low

  * Add missing information in debian/control
  * Add missing build dependency

 -- Bruno Araujo <bruno.araujo@openbossa.org>  Mon, 27 Jun 2011 14:17:12 -0400

pyside-assistant (0.1-1) unstable; urgency=low

  * Initial release

 -- Bruno Araujo <bruno.araujo@openbossa.org>  Thu, 05 May 2011 11:20:02 -0400


tests/arfile.py

'''Helper module to deal with arfiles (e.g. deb files)'''

import os
import sys


def is_arfile(archive):
    with open(archive, 'rb') as handle:
        return handle.read(8) == "!<arch>\n"


def get_members(archive):
    '''Gets a list of all members of this archive'''

    if not is_arfile(archive):
        return []

    members = []

    with open(archive, 'rb') as handle:
        handle.read(8)

        while True:
            header = handle.read(60)
            if not header.strip():
                break
            filename = header[:16].strip()
            if filename.endswith('/'): # Strip GNU extensions
                filename = filename[:-1]
            members.append(filename)
            filesize = int(header[48:58])
            handle.seek(filesize, os.SEEK_CUR)

    return members


def extract(archive, name=None, targetdir=None):
    '''Extract all files or just 'name' from 'archive' into targetdir'''

    if targetdir is None:
        targetdir = os.curdir

    if not is_arfile(archive):
        raise ValueError('Not an ar file')

    with open(archive, 'rb') as handle:
        handle.read(8)

        while True:
            header = handle.read(60)
            if not header.strip():
                break

            filename = header[:16].strip()
            if filename.endswith('/'):
                filename = filename[:-1]

            filesize = int(header[48:58])

            if not name or filename == name:
                with open(os.path.join(targetdir, filename), 'wb') as output:
                    filedata = handle.read(filesize)
                    output.write(filedata)
                    if filesize % 2:
                        handle.read(1)
                    #TODO handle permissions?
                if name:
                    break
            else:
                if filesize % 2:
                    filesize += 1
                handle.seek(filesize, os.SEEK_CUR)


if __name__ == '__main__':
    name = sys.argv[1]
    print is_arfile(name)

    print get_members(name)

    extract(name)



tests/psa_commands_test.py
'''Unit tests for pyside-assistant commands'''

import unittest
import shutil
import os
import subprocess
import sys
import tempfile
from contextlib import contextmanager

import tarfile
import arfile

@contextmanager
def working_directory(path):
    '''Simple context manager to change the working directory'''
    current_dir = os.getcwd()
    os.chdir(path)
    try:
        yield
    finally:
        os.chdir(current_dir)


class PySideAssistantCommandsTest(unittest.TestCase):

    def setUp(self):
        self.path = tempfile.mkdtemp(prefix='psatemp')
        shutil.rmtree(self.path, ignore_errors=True)
        os.makedirs(self.path)

    def tearDown(self):
        shutil.rmtree(self.path)

    def runShellCommand(self, command, verbose=False):
        proc = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE,
                                stderr=subprocess.PIPE)
        stdout, stderr = proc.communicate()
        if proc.returncode:
            self.fail(stderr)

    def verifyDirectoryStructure(self, root, files):
        for filename in files:
            self.assert_(os.path.exists(os.path.join(root, filename)), msg='File %s not found' % filename)


class InitTest(PySideAssistantCommandsTest):
    def testInitCommandHarmattan(self):
        with working_directory(self.path):
            command = 'psa init testproject-harmattan harmattan> /dev/null'
            self.runShellCommand(command)

        #assert all files are in place
        project_path = os.path.join(self.path, 'testproject-harmattan')

        filenames = [
         'testproject-harmattan.psa',
         'testproject-harmattan.aegis',
         'testproject-harmattan',
         'MANIFEST.in',
         'setup.py',
         'testproject-harmattan.longdesc',
         'stdeb.cfg',
         'testproject-harmattan.desktop',
         'qml',
         'qml/MainPage.qml',
         'qml/main.qml',
         'testproject-harmattan.png',
         'README.assistant',
        ]

        self.verifyDirectoryStructure(project_path, filenames)

    def testInitCommandFremantle(self):
        with working_directory(self.path):
            command = 'psa init testproject-fremantle fremantle > /dev/null'
            self.runShellCommand(command)

        #assert all files are in place
        project_path = os.path.join(self.path, 'testproject-fremantle')

        filenames = [
         'testproject-fremantle.psa',
         'testproject-fremantle',
         'MANIFEST.in',
         'setup.py',
         'testproject-fremantle.longdesc',
         'stdeb.cfg',
         'testproject-fremantle.desktop',
         'qml',
         'qml/main.qml',
         'testproject-fremantle.png',
         'README.assistant',
        ]

        self.verifyDirectoryStructure(project_path, filenames)

    def testInitCommandUbuntu(self):
        with working_directory(self.path):
            command = 'psa init testproject-ubuntu ubuntu-qml > /dev/null'
            self.runShellCommand(command)

        #assert all files are in place
        project_path = os.path.join(self.path, 'testproject-ubuntu')

        filenames = [
         'testproject-ubuntu.psa',
         'testproject-ubuntu',
         'MANIFEST.in',
         'setup.py',
         'testproject-ubuntu.longdesc',
         'stdeb.cfg',
         'testproject-ubuntu.desktop',
         'qml',
         'qml/main.qml',
         'testproject-ubuntu.png',
         'README.assistant',
        ]

        self.verifyDirectoryStructure(project_path, filenames)

    def testInitCommandUbuntu(self):
        with working_directory(self.path):
            command = 'psa init testproject-ubuntu-gui ubuntu-qtgui > /dev/null'
            self.runShellCommand(command)

        #assert all files are in place
        project_path = os.path.join(self.path, 'testproject-ubuntu-gui')

        filenames = [
         'testproject-ubuntu-gui.psa',
         'testproject-ubuntu-gui',
         'MANIFEST.in',
         'setup.py',
         'testproject-ubuntu-gui.longdesc',
         'stdeb.cfg',
         'testproject-ubuntu-gui.desktop',
         'testproject-ubuntu-gui.png',
         'README.assistant',
        ]

        self.verifyDirectoryStructure(project_path, filenames)

    def testInitCommandParameters(self):
        #test long commands
        with working_directory(self.path):
            command = 'psa init testproject1 harmattan --section="games" --app-name="test app1" --category="Game" --description="a description1" > /dev/null'
            self.runShellCommand(command)

        f = open(os.path.join(self.path, 'testproject1', 'stdeb.cfg'))
        contents = f.read()
        f.close()
        self.assert_('Section: user/games' in contents)

        f = open(os.path.join(self.path, 'testproject1', 'testproject1.desktop'))
        contents = f.read()
        f.close()
        self.assert_('Name=test app1' in contents)
        self.assert_('Categories=Game;' in contents)

        f = open(os.path.join(self.path, 'testproject1', 'setup.py'))
        contents = f.read()
        f.close()
        self.assert_('description="a description1"' in contents)

        #test short commands
        command = ' '.join(['cd', self.path, ';', 'psa init testproject2 harmattan -s "network" -a "test app2" -c "Video" -d "a description2" > /dev/null'])
        self.runShellCommand(command)

        f = open(os.path.join(self.path, 'testproject2', 'stdeb.cfg'))
        contents = f.read()
        f.close()
        self.assert_('Section: user/network' in contents)

        f = open(os.path.join(self.path, 'testproject2', 'testproject2.desktop'))
        contents = f.read()
        f.close()
        self.assert_('Name=test app2' in contents)
        self.assert_('Categories=Video;' in contents)

        f = open(os.path.join(self.path, 'testproject2', 'setup.py'))
        contents = f.read()
        f.close()
        self.assert_('description="a description2"' in contents)


class BuildTest(PySideAssistantCommandsTest):


    def init_project(self, project, templatename):
        with working_directory(self.path):
            command = 'psa init %s %s > /dev/null' % (project, templatename)
            self.runShellCommand(command)

        return os.path.join(self.path, project)

    def build_deb(self, project, path):
        expected_deb = os.path.join(path, 'deb_dist', ('%s_0.1.0-1_all.deb' % project))
        with working_directory(os.path.join(self.path, project)):
            command = 'psa build-deb > /dev/null'
            self.runShellCommand(command)
        self.assert_(os.path.exists(expected_deb), msg="Debian file %s does not exist" % expected_deb)

        return expected_deb

    def check_deb_contents(self, deb, deb_contents):
        path = tempfile.mkdtemp(prefix='psa_deb')

        try:
            arfile.extract(deb, targetdir=path)

            rootfiles = deb_contents['root']
            for filename in rootfiles:
                self.assertTrue(os.path.exists(os.path.join(path, filename)))

            # control.tar.gz
            files = deb_contents['control']
            tar = tarfile.open(os.path.join(path, 'control.tar.gz'), 'r')

            try:
                for filename in files:
                    self.assertTrue(tar.getmember(filename))
            finally:
                tar.close()

            # data.tar.gz
            files = deb_contents['data']
            tar = tarfile.open(os.path.join(path, 'data.tar.gz'), 'r')

            try:
                for filename in files:
                    self.assertTrue(tar.getmember(filename))
            finally:
                tar.close()
        finally:
            shutil.rmtree(path)

    def base_debian_components(self):
        '''Base debian components common to all packages'''

        return {
                'root': ['debian-binary'],
                'control': [
                        './control',
                        './md5sums',
                        './postinst',
                ],
                'data': [
                ]
        }

    def testBuildHarmattan(self):

        project = 'foobar'

        path = self.init_project(project, 'harmattan')

        with open(os.path.join(path, project+'.aegis'), 'w') as handle:
            handle.write('The quick brown fox jumps over the lazy dog')

        deb = self.build_deb(project, path)

        deb_contents = self.base_debian_components()

        deb_contents['root'].append('./_aegis')
        deb_contents['control'].append('./digsigsums')
        deb_contents['data'].append('./usr/bin/%s' % project)
        deb_contents['data'].append('./usr/share/applications/%s.desktop' % project)
        deb_contents['data'].append('./usr/share/icons/hicolor/64x64/apps/%s.png' % project)
        deb_contents['data'].append('./usr/share/%s/qml/main.qml' % project)
        deb_contents['data'].append('./usr/share/%s/qml/MainPage.qml' % project)

        self.check_deb_contents(deb, deb_contents)

    def testBuildFremantle(self):
        project = 'foobar'

        path = self.init_project(project, 'fremantle')

        deb = self.build_deb(project, path)

        deb_contents = self.base_debian_components()

        deb_contents['data'].append('./usr/bin/%s' % project)
        deb_contents['data'].append('./usr/share/applications/hildon/%s.desktop' % project)
        deb_contents['data'].append('./usr/share/icons/%s.png' % project)
        deb_contents['data'].append('./opt/usr/share/%s/qml/main.qml' % project)

        self.check_deb_contents(deb, deb_contents)

    def testBuildUbuntu(self):
        project = 'foobar'

        path = self.init_project(project, 'ubuntu-qml')

        deb = self.build_deb(project, path)

        deb_contents = self.base_debian_components()

        deb_contents['data'].append('./usr/bin/%s' % project)
        deb_contents['data'].append('./usr/share/applications/%s.desktop' % project)
        deb_contents['data'].append('./usr/share/pixmaps/%s.png' % project)
        deb_contents['data'].append('./usr/share/%s/qml/main.qml' % project)

        self.check_deb_contents(deb, deb_contents)

    def testBuildUbuntuGui(self):
        project = 'foobar'

        path = self.init_project(project, 'ubuntu-qtgui')

        deb = self.build_deb(project, path)

        deb_contents = self.base_debian_components()

        deb_contents['data'].append('./usr/bin/%s' % project)
        deb_contents['data'].append('./usr/share/applications/%s.desktop' % project)
        deb_contents['data'].append('./usr/share/pixmaps/%s.png' % project)

        self.check_deb_contents(deb, deb_contents)

class UpdateTest(PySideAssistantCommandsTest):

    def testUpdateCommand(self):
        command = ' '.join(['cd', self.path, ';', 'psa init testproject harmattan > /dev/null'])
        self.runShellCommand(command)

        #test long commands
        command = ' '.join(['cd', os.path.join(self.path,'testproject'), ';', 'psa update --section="games" --app-name="test app1" --category="Game" --description="a description1" > /dev/null'])
        self.runShellCommand(command)
        f = open(os.path.join(self.path, 'testproject', 'stdeb.cfg'))
        contents = f.read()
        f.close()
        self.assert_('Section: user/games' in contents)

        f = open(os.path.join(self.path, 'testproject', 'testproject.desktop'))
        contents = f.read()
        f.close()
        self.assert_('Name=test app1' in contents)
        self.assert_('Categories=Game;' in contents)

        f = open(os.path.join(self.path, 'testproject', 'setup.py'))
        contents = f.read()
        f.close()
        self.assert_('description="a description1"' in contents)

        #test short commands
        command = ' '.join(['cd', os.path.join(self.path,'testproject'), ';', 'psa update -s "network" -a "test app2" -c "Video" -d "a description2" > /dev/null'])
        self.runShellCommand(command)

        f = open(os.path.join(self.path, 'testproject', 'stdeb.cfg'))
        contents = f.read()
        f.close()
        self.assert_('Section: user/network' in contents)

        f = open(os.path.join(self.path, 'testproject', 'testproject.desktop'))
        contents = f.read()
        f.close()
        self.assert_('Name=test app2' in contents)
        self.assert_('Categories=Video;' in contents)

        f = open(os.path.join(self.path, 'testproject', 'setup.py'))
        contents = f.read()
        f.close()
        self.assert_('description="a description2"' in contents)


if __name__ == "__main__":
    unittest.main()


scripts/deb-add.py
#!/usr/bin/python
# This file is part of the PySide project.
#
# Copyright (C) 2011 Nokia Corporation and/or its subsidiary(-ies).
#
# Contact: PySide team <contact@pyside.org>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# version 2 as published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
# 02110-1301 USA

''' deb-add - Python port of the Perl script with a similar name that
adds/replaces files in binary debian packages.'''

import sys
import os
import re
import logging
import stat
from optparse import OptionParser

def parse_args():
    '''Parse command line options from sys.argv'''
    parser = OptionParser()

    parser.add_option('-c', '--control', dest='control',
                      help='Control file for reading.',
                      action='store', type='string')

    parser.set_defaults(control='', verbose=False)

    return parser.parse_args()

def parse_control(control):
    '''Get name, version and arch from a debian control file'''
    name, version, arch = '', '', ''
    namere = re.compile('^Package:\s*(\S+)\s*$')
    versionre = re.compile('^Version:\s*(\S+)\s*$')
    archre = re.compile('^Architecture:\s*(\S+)\s*$')
    with open(control) as handle:
        for line in handle:
            match = namere.search(line)
            if match:
                name = match.groups()[0]
                continue
            match = versionre.search(line)
            if match:
                version = match.groups()[0]
                continue
            match = archre.search(line)
            if match:
                arch = match.groups()[0]
                continue

    logging.debug('Found name %s, version %s, arch %s', (name, version, arch))
    if not all([name, version, arch]):
        logging.critical("Couldn't parse control file %s", control)
        sys.exit(1)

    return name, version, arch


def replace_files(newdeb, debfile, changes, control):
    '''Replace or copy the existing files in the debian archive'''
    name = control[0]
    to_remove = []

    with open(debfile, 'rb') as deb:

        # magic:
        magic = '!<arch>\n'
        debmagic = deb.read(8)

        if debmagic != magic:
            logging.critical("File %s does not have .deb magic number", debfile)
            sys.exit(1)

        newdeb.write(magic)

        # ar_name[16];    +00
        # ar_date[12];    +16 (= seconds since)
        # ar_uid[6]       +28 (= "0     ")
        # ar_gid[6]       +34 (= "0     ")
        # ar_mode[8]      +40 (= "100644  ")
        # ar_size[10];    +48
        # ar_fmag[2];     +58 (= "`\n")
        # -------------------
        #                 =60

        # List existing files that could be replaced.
        while 1:
            header = deb.read(60)
            if len(header) <= 0:
                break
            member = header[:16]
            size = int(header[48:58])
            if size & 1:
                size += 1

            if header[58:60] != '`\n':
                logging.warning('Bad AR header.')
                break # We're done with the existing files.

            # Should we replace it?
            if member.strip() in changes:
                source, mtime, newsize, target = changes[member.strip()]
                to_remove.append(member.strip())

                print 'Replacing', member

                header = header[:16] + '%-12s' % mtime + header[16+12:]
                header = header[:48] + '%-10s' % newsize + header[48+10:]
                newdeb.write(header)
                with open(source, 'rb') as temp:
                    buf = temp.read(newsize)
                    if len(buf) != newsize:
                        logging.critical('Failed to read %s fully', source)
                        sys.exit(0)
                    newdeb.write(buf)
                if newsize & 1:
                    newdeb.write('\n')

            newdeb.write(header)
            newdeb.write(deb.read(size))

    # Update files dict to avoid files being added again.
    for name in to_remove:
        del changes[name]


def add_files(newdeb, files):
    '''Add new files to the debian archive.'''

    for name, info in files.items():
        print 'Inserting', name
        if not info[3]:
            continue
        newsize = info[2]
        newhdr = "%-16s%-12s%-6s%-6s%-8s%-10s`\n"
        newhdr %= (info[3],
                        info[1],
                        '0',
                        '0',
                        '100644',
                        info[2])

        newdeb.write(newhdr)
        with open(info[0], 'rb') as temp:
            buf = temp.read(newsize)
            if len(buf) != newsize:
                logging.critical('Failed to read %s fully', info[0])
                sys.exit(1)
            newdeb.write(buf)
        if newsize & 1:
            newdeb.write('\n')

def process_args(args):
    '''Process the arguments with the file pairs
    Return a dict of tuples.

    Key: target file in the deb archive
    Values: (source file, last modification time, size in bytes, target file.
    '''

    files = {}

    for arg in args:
        try:
            source, target = arg.split('=')
        except ValueError:
            logging.critical('Filenames must be <source>=<target>')
            sys.exit(1)

        if not (os.path.exists(source) and os.path.isfile(source)):
            logging.critical('Source file must be a regular file')
            sys.exit(1)

        if len(target) > 16:
            logging.critical('Target file must be at most 16 characters')
            sys.exit(1)

        st = os.stat(source)

        files[target] = (source, st[stat.ST_MTIME], st[stat.ST_SIZE], target)

    return files


def main():

    options, args = parse_args()

    if options.verbose:
        logging.basicConfig(level=logging.DEBUG)

    if options.control:
        # try to parse the provided control file
        name, version, arch = parse_control(options.control)

        control = name, version, arch
        debfile = '%s_%s_%s.deb' % (name, version, arch)
    else:
        if not args:
            logging.critical('Must provide a control or debian file.')
            sys.exit(1)
        debfile = args[0]

    files = process_args(args)

    if not files:
        logging.warning('No files to be injected. Exiting')
        sys.exit(0)

    with open(debfile+'.new', 'wb') as newdeb:

        replace_files(newdeb, debfile, files, control)

        add_files(newdeb, files)

    os.rename(debfile+'.new', debfile)


if __name__ == '__main__':
    main()


scripts/deb_add.py
#!/usr/bin/python
# This file is part of the PySide project.
#
# Copyright (C) 2011 Nokia Corporation and/or its subsidiary(-ies).
#
# Contact: PySide team <contact@pyside.org>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# version 2 as published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
# 02110-1301 USA

''' deb-add - Python port of the Perl script with a similar name that
adds/replaces files in binary debian packages.'''

import sys
import os
import re
import logging
import stat
from optparse import OptionParser

def parse_args():
    '''Parse command line options from sys.argv'''
    parser = OptionParser()

    parser.add_option('-c', '--control', dest='control',
                      help='Control file for reading.',
                      action='store', type='string')

    parser.set_defaults(control='', verbose=False)

    return parser.parse_args()

def parse_control(control):
    '''Get name, version and arch from a debian control file'''
    name, version, arch = '', '', ''
    namere = re.compile('^Package:\s*(\S+)\s*$')
    versionre = re.compile('^Version:\s*(\S+)\s*$')
    archre = re.compile('^Architecture:\s*(\S+)\s*$')
    with open(control) as handle:
        for line in handle:
            match = namere.search(line)
            if match:
                name = match.groups()[0]
                continue
            match = versionre.search(line)
            if match:
                version = match.groups()[0]
                continue
            match = archre.search(line)
            if match:
                arch = match.groups()[0]
                continue

    logging.debug('Found name %s, version %s, arch %s', name, version, arch)
    if not all([name, version, arch]):
        logging.critical("Couldn't parse control file %s", control)
        sys.exit(1)

    return name, version, arch


def replace_files(newdeb, debfile, changes, control):
    '''Replace or copy the existing files in the debian archive'''
    name = control[0]
    to_remove = []

    with open(debfile, 'rb') as deb:

        # magic:
        magic = '!<arch>\n'
        debmagic = deb.read(8)

        if debmagic != magic:
            logging.critical("File %s does not have .deb magic number", debfile)
            sys.exit(1)

        newdeb.write(magic)

        # ar_name[16];    +00
        # ar_date[12];    +16 (= seconds since)
        # ar_uid[6]       +28 (= "0     ")
        # ar_gid[6]       +34 (= "0     ")
        # ar_mode[8]      +40 (= "100644  ")
        # ar_size[10];    +48
        # ar_fmag[2];     +58 (= "`\n")
        # -------------------
        #                 =60

        # List existing files that could be replaced.
        while 1:
            header = deb.read(60)
            if len(header) <= 0:
                break
            member = header[:16]
            size = int(header[48:58])
            if size & 1:
                size += 1

            if header[58:60] != '`\n':
                logging.warning('Bad AR header.')
                break # We're done with the existing files.

            # Should we replace it?
            if member.strip() in changes:
                source, mtime, newsize, target = changes[member.strip()]
                to_remove.append(member.strip())

                print 'Replacing', member

                header = header[:16] + '%-12s' % mtime + header[16+12:]
                header = header[:48] + '%-10s' % newsize + header[48+10:]
                newdeb.write(header)
                with open(source, 'rb') as temp:
                    buf = temp.read(newsize)
                    if len(buf) != newsize:
                        logging.critical('Failed to read %s fully', source)
                        sys.exit(0)
                    newdeb.write(buf)
                if newsize & 1:
                    newdeb.write('\n')

            newdeb.write(header)
            newdeb.write(deb.read(size))

    # Update files dict to avoid files being added again.
    for name in to_remove:
        del changes[name]


def add_files(newdeb, files):
    '''Add new files to the debian archive.'''

    for name, info in files.items():
        print 'Inserting', name
        if not info[3]:
            continue
        newsize = info[2]
        newhdr = "%-16s%-12s%-6s%-6s%-8s%-10s`\n"
        newhdr %= (info[3],
                        info[1],
                        '0',
                        '0',
                        '100644',
                        info[2])

        newdeb.write(newhdr)
        with open(info[0], 'rb') as temp:
            buf = temp.read(newsize)
            if len(buf) != newsize:
                logging.critical('Failed to read %s fully', info[0])
                sys.exit(1)
            newdeb.write(buf)
        if newsize & 1:
            newdeb.write('\n')

def process_args(args):
    '''Process the arguments with the file pairs
    Return a dict of tuples.

    Key: target file in the deb archive
    Values: (source file, last modification time, size in bytes, target file.
    '''

    files = {}

    for arg in args:
        try:
            source, target = arg.split('=')
        except ValueError:
            logging.critical('Filenames must be <source>=<target>')
            sys.exit(1)

        if not (os.path.exists(source) and os.path.isfile(source)):
            logging.critical('Source file must be a regular file')
            sys.exit(1)

        if len(target) > 16:
            logging.critical('Target file must be at most 16 characters')
            sys.exit(1)

        st = os.stat(source)

        files[target] = (source, st[stat.ST_MTIME], st[stat.ST_SIZE], target)

    return files


def validate_file(abs_source, target):
    '''Validates the source, target pair'''

    logging.info('Validating source %s and target %s', abs_source, target)
    if not (os.path.exists(abs_source) and os.path.isfile(abs_source)):
        raise ValueError('Source file must be a regular file')

    if len(target) > 16:
        raise ValueError('Target file must be at most 16 characters')


def add(abs_source, target, abs_control, abs_debfile):
    '''Adds file to a existing debian package.

    All paths are absolute, except for target which is relative to
    the package root.

    abs_source - absolute path of the source file
    target - target name of the file in the package
    abs_control - control file of this package
    abs_debfile - debian package to be modified.

    This function copies the original debian package to abs_debfile.orig
    '''

    logging.info('Adding %s as %s to abs_debfile %s with abs_control %s', abs_source, target, abs_debfile, abs_control)

    control_data = parse_control(abs_control)
    validate_file(abs_source, target)

    st = os.stat(abs_source)

    files = {target : (abs_source, st[stat.ST_MTIME], st[stat.ST_SIZE], target)}

    with open(abs_debfile+'.new', 'wb') as newdeb:
        replace_files(newdeb, abs_debfile, files, control_data)
        add_files(newdeb, files)

    os.rename(abs_debfile, abs_debfile + '.orig')
    os.rename(abs_debfile + '.new', abs_debfile)


def main():

    options, args = parse_args()

    if options.verbose:
        logging.basicConfig(level=logging.DEBUG)

    if options.control:
        name, version, arch = parse_control(options.control)

        control = name, version, arch
        debfile = '%s_%s_%s.deb' % (name, version, arch)
    else:
        if not args:
            logging.critical('Must provide a control or debian file.')
            sys.exit(1)
        debfile = args[0]

    files = process_args(args)

    if not files:
        logging.warning('No files to be injected. Exiting')
        sys.exit(0)

    with open(debfile+'.new', 'wb') as newdeb:

        replace_files(newdeb, debfile, files, control)

        add_files(newdeb, files)

    os.rename(debfile+'.new', debfile)


if __name__ == '__main__':
    main()
